New Capture analysis

arora-aman opened the pull request:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/sexxi-goose/rust/28)
<!-- Reviewable:end -->

arora-aman committed:
Use Places to express closure/generator Captures

Co-authored-by: Archer Zhang <archer.xn@gmail.com>

arora-aman committed:
Initial work for doing minimum capture analysis for RFC-2229

Co-authored-by: Chris Pardy <chrispardy36@gmail.com>
Co-authored-by: Logan Mosier <logmosier@gmail.com>

arora-aman committed:
Indroduce feature flag for RFC-2229

Signed-off-by: Aman Arora <me@aman-arora.com>

arora-aman committed:
Add initial set of testcases for RFC 2229

Co-authored-by: Dhruv Jauhar <dhruvjhr@gmail.com>

arora-aman committed:
Update tests with min capture information

Co-authored-by: Chris Pardy <chrispardy36@gmail.com>
Co-authored-by: Logan Mosier <logmosier@gmail.com>

arora-aman committed:
Add helper function for Capture Esclations and expressions

Co-authored-by: Dhruv Jauhar <dhruvjhr@gmail.com>

arora-aman commented on compiler/rustc_typeck/src/check/upvar.rs:
```diff
@@ -522,10 +885,60 @@ impl<'a, 'tcx> euv::Delegate<'tcx> for InferBorrowKind<'a, 'tcx> {
+
+    let mut same_initial_projections = true;
+
+    // Returns (equal, array_equal)
```
Remove this commment

arora-aman commented on compiler/rustc_typeck/src/check/upvar.rs:
```diff
@@ -250,6 +274,326 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
+        // expressions.
+        let eq_capture_kind = match (capture_info_a.capture_kind, capture_info_b.capture_kind) {
+            (ty::UpvarCapture::ByValue(_), ty::UpvarCapture::ByValue(_)) => {
+                // Both are either Some or both are either None
```
Remove these lines

arora-aman commented on compiler/rustc_middle/src/ty/mod.rs:
```diff
@@ -752,9 +759,44 @@ pub struct UpvarBorrow<'tcx> {
 
+#[derive(PartialEq, Clone, Debug, Copy, TyEncodable, TyDecodable, HashStable)]
+pub struct CaptureInfo<'tcx> {
+    /// Expr Id pointing to use that resulting in selecting the current capture kind
```
resulted in selecting

arora-aman commented on compiler/rustc_typeck/src/expr_use_visitor.rs:
```diff
@@ -536,55 +537,91 @@ impl<'a, 'tcx> ExprUseVisitor<'a, 'tcx> {
                         ty::BindByValue(..) => {
                             let mode = copy_or_move(mc, place);
+
+                            // Pointing to patterns in diagnostics can cause confusion,
```
This is a hack right now, will be fixed by https://github.com/rust-lang/project-rfc-2229/issues/20

arora-aman commented on compiler/rustc_typeck/Cargo.toml:
```diff
@@ -9,6 +9,7 @@ test = false
 doctest = false
 
 [dependencies]
+indexmap = "1.5.1"
```
Do we need this 🤔

arora-aman commented on compiler/rustc_typeck/src/expr_use_visitor.rs:
```diff
@@ -536,55 +537,91 @@ impl<'a, 'tcx> ExprUseVisitor<'a, 'tcx> {
+    /// When the current body being handled is a closure, then we must make sure that
+    /// - The parent closure only captures Places from the nested closure that are not local to it.
+    ///
+    /// In the following example the closures `c` cares about only `p.x` and not `incr`.
```
should only capture `p.x` even though `incr` is a capture of the nested closure

roxelo commented on compiler/rustc_middle/src/ty/mod.rs:
````diff
@@ -752,9 +759,40 @@ pub struct UpvarBorrow<'tcx> {
+/// Consider closure where s.str1 is captured via an ImmutableBorrow and s.str2 via a MutableBorrow
+///
+/// ```rust
+/// // Assume that thte HirId for the variable definition is `V1`
````
spelling

roxelo commented on compiler/rustc_typeck/src/check/upvar.rs:
```diff
@@ -284,30 +284,17 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
-            }
+                // upvar_capture_map only stores the UpvarCapture (CaptureKind),
+                // so we create a fake capture info with no expression.
+                let fake_capture_info =
```
fake capture info? I am kind of confused. So we choose between two types of different capture based on `determine_capture_info`, so is it really a fake_capture_info? Do you mean it's the old capture info?

arora-aman replied:
The reason it's fake is because it's not the actual capture_info that resulted in the existing entry in the map, espcially note the None expr_id

roxelo commented on compiler/rustc_typeck/src/check/upvar.rs:
```diff
@@ -522,42 +565,28 @@ impl<'a, 'tcx> InferBorrowKind<'a, 'tcx> {
-            }
-        }
+        if let ty::UpvarCapture::ByValue(_) = curr_capture_info.capture_kind {
+            // It's already capture by value, we don't need to do anything here
```
captured?

roxelo commented on compiler/rustc_middle/src/ty/context.rs:
```diff
@@ -415,9 +415,10 @@ pub struct TypeckResults<'tcx> {
-    /// `Place`s and how/why are they captured by the closure.
-    pub closure_capture_information: ty::CaptureInformationMap<'tcx>,
+    /// Given the closure DefId this map provides a map of
+    /// root variarbles to minimum set of `Place`s (and how) that need to be tracked
```
variables

roxelo commented on compiler/rustc_mir/src/borrow_check/diagnostics/mod.rs:
```diff
@@ -967,9 +964,15 @@ impl<'cx, 'tcx> MirBorrowckCtxt<'cx, 'tcx> {
+                    .zip(places)
             {
+                // upvars_mentioned stores var_hir_id -> span
+                // closure captures, for give closure var_hir_id -> UpvarId=(var_hir_id, local def
```
for give closure? Not sure I understand what you mean here..

arora-aman replied:
I'll remove this comment, doesn't really need to be here. But essentially for a give closure c, closure_captures[c] gives us a map from var_hir_id -> UpvarId

roxelo commented on compiler/rustc_typeck/src/check/upvar.rs:
```diff
@@ -433,8 +618,27 @@ struct InferBorrowKind<'a, 'tcx> {
-    // For each upvar that we access, we track the minimal kind of
-    // access we need (ref, ref mut, move, etc).
+    /// For each Place that we access, we track the minimal kind of
+    /// access we need (ref, ref mut, move, etc) and the expressiion that resulted in such access.
```
expression

roxelo commented on compiler/rustc_typeck/src/check/upvar.rs:
````diff
@@ -433,8 +618,27 @@ struct InferBorrowKind<'a, 'tcx> {
+    /// s.str2 via a MutableBorrow
+    ///
+    /// ```rust
+    /// // Assume that thte HirId for the variable definition is `V1`
````
the

roxelo commented on compiler/rustc_typeck/src/expr_use_visitor.rs:
````diff
@@ -555,47 +554,69 @@ impl<'a, 'tcx> ExprUseVisitor<'a, 'tcx> {
+    /// }
+    /// ```
+    ///
+    /// - When reporting the Place back to the Delgate, ensure that the UpvarId uses the enclosing
````
Delegate?

roxelo reviewed (commented):
Mainly nit for now. Will take a second read tomorrow with a bigger focus on the code itself

roxelo committed:
Reduce verbosity of capture analysis logs

Co-authored-by: Jenny Wills <wills.jenniferg@gmail.com>
Co-authored-by: Aman Arora <me@aman-arora.com>

roxelo commented on compiler/rustc_typeck/src/check/upvar.rs:
```diff
@@ -39,10 +39,37 @@ use rustc_hir::def_id::DefId;
+
+use std::env;
+
+macro_rules! log_capture_analysis {
+    ($fcx:expr, $closure_def_id:expr, $fmt:literal) => {
+        if $fcx.should_log_capture_analysis($closure_def_id) {
+            debug!("For closure={:?}: {}", $closure_def_id, $fmt);
+        }
+    };
+
+    ($fcx:expr, $closure_def_id:expr, $fmt:literal, $($args:expr),*) => {
+        if $fcx.should_log_capture_analysis($closure_def_id) {
+            debug!(concat!("For closure={:?}: ", $fmt), $closure_def_id, $($args),*);
+        }
+    };
+}
```
Should we keep this macro? Maybe let's just remain consistent and move everything to aprinting method?

arora-aman replied:
Updated the last commit

roxelo reviewed (approved):
LGTM

Azhng reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/sexxi-goose/rust/28)* status: 0 of 33 files reviewed, 17 unresolved discussions (waiting on @arora-aman and @roxelo)

___
*[compiler/rustc_mir/src/borrow_check/diagnostics/mod.rs, line 386 at r5](https://reviewable.io/reviews/sexxi-goose/rust/28#-ML-bHgb5V7kQh5pfti3:-ML-bHgb5V7kQh5pfti4:bqlo17j) ([raw file](https://github.com/sexxi-goose/rust/blob/ab4b8ce9ab15b1a6d58c6aa7a3ceb952d6206814/compiler/rustc_mir/src/borrow_check/diagnostics/mod.rs#L386)):*
> ```Rust
>                 }
>                 ty::Closure(def_id, _) | ty::Generator(def_id, _, _) => {
>                     // `tcx.upvars_mentioned(def_id)` returns an `Option`, which is `None` in case
> ```

The comment here need to be updated since we are reading from `closure_captures` instead of `upvars_mentioned`
___
*[compiler/rustc_typeck/src/expr_use_visitor.rs, line 596 at r5](https://reviewable.io/reviews/sexxi-goose/rust/28#-ML-eIwG1gDHIIxfMWaZ:-ML-eIwG1gDHIIxfMWa_:blltxsq) ([raw file](https://github.com/sexxi-goose/rust/blob/ab4b8ce9ab15b1a6d58c6aa7a3ceb952d6206814/compiler/rustc_typeck/src/expr_use_visitor.rs#L596)):*
> ```Rust
>                     if !upvars
>                         .map_or(!body_owner_is_closure, |upvars| upvars.contains_key(var_hir_id))
>                     {
> ```

Move this if check to the outer loops? Seems like we are doing redundant checks in the inner tight loops since `var_hir_id` is not updated in the inner loop

Also this double negative seems really strange. Not sure if there's a better way of writing it 🤷‍♀️🤷‍♀️🤷‍♀️
___
*[compiler/rustc_typeck/src/check/upvar.rs, line 53 at r5](https://reviewable.io/reviews/sexxi-goose/rust/28#-ML-f0p79ZtaSF2ZIbx7:-ML-f0p79ZtaSF2ZIbx8:bs34w4t) ([raw file](https://github.com/sexxi-goose/rust/blob/ab4b8ce9ab15b1a6d58c6aa7a3ceb952d6206814/compiler/rustc_typeck/src/check/upvar.rs#L53)):*
> ```Rust
> /// - foo.bar.baz is an descendant of foo.bar,
> /// - foo is ancestor of foo.bar.baz
> /// - foo.bar and foo.baz are divergent
> ```

nit: the order of comment is inconsistent with the definition of the enum 
___
*[compiler/rustc_typeck/src/check/upvar.rs, line 130 at r5](https://reviewable.io/reviews/sexxi-goose/rust/28#-ML-fh9a8ugBdOs6qPIu:-ML-fh9a8ugBdOs6qPIv:b-xo6ott) ([raw file](https://github.com/sexxi-goose/rust/blob/ab4b8ce9ab15b1a6d58c6aa7a3ceb952d6206814/compiler/rustc_typeck/src/check/upvar.rs#L130)):*
> ```Rust
> 
>         let mut capture_information = FxIndexMap::<Place<'tcx>, ty::CaptureInfo<'tcx>>::default();
>         if self.tcx.features().capture_disjoint_fields || matches!(env::var("SG_NEW"), Ok(_)) {
> ```

So we are keeping the `SG_NEW` env var ?
___
*[compiler/rustc_typeck/src/check/upvar.rs, line 153 at r5](https://reviewable.io/reviews/sexxi-goose/rust/28#-ML-hD9t7Wadq6wo5Qsa:-ML-hD9t7Wadq6wo5Qsb:b-ef8376) ([raw file](https://github.com/sexxi-goose/rust/blob/ab4b8ce9ab15b1a6d58c6aa7a3ceb952d6206814/compiler/rustc_typeck/src/check/upvar.rs#L153)):*
> ```Rust
>             closure_def_id,
>             closure_span: span,
>             capture_clause,
> ```

nit: maybe specified the name of the field name that is being initialized, it's a mix and match here and looks kinda inconsistent
___
*[compiler/rustc_typeck/src/check/upvar.rs, line 156 at r5](https://reviewable.io/reviews/sexxi-goose/rust/28#-ML-hMcnB4PAsMIAX-vY:-ML-hMcnB4PAsMIAX-vZ:b-rqxvol) ([raw file](https://github.com/sexxi-goose/rust/blob/ab4b8ce9ab15b1a6d58c6aa7a3ceb952d6206814/compiler/rustc_typeck/src/check/upvar.rs#L156)):*
> ```Rust
>             current_closure_kind: ty::ClosureKind::LATTICE_BOTTOM,
>             current_origin: None,
>             capture_information,
> ```

ditto 
___
*[compiler/rustc_typeck/src/check/upvar.rs, line 345 at r5](https://reviewable.io/reviews/sexxi-goose/rust/28#-ML-k8hc8BHofKaVCaEA:-ML-k8hc8BHofKaVCaEB:bht1lll) ([raw file](https://github.com/sexxi-goose/rust/blob/ab4b8ce9ab15b1a6d58c6aa7a3ceb952d6206814/compiler/rustc_typeck/src/check/upvar.rs#L345)):*
> ```Rust
>     /// ```rust
>     /// let s: String;  // hir_id_s
>     /// let mut p: Point; // his_id_p
> ```

Assuming `Point` only has field `x` and `y` here ? 
___
*[compiler/rustc_typeck/src/check/upvar.rs, line 361 at r5](https://reviewable.io/reviews/sexxi-goose/rust/28#-ML-kJvn8u6FjJgGXENE:-ML-kJvn8u6FjJgGXENF:b-m3rr57) ([raw file](https://github.com/sexxi-goose/rust/blob/ab4b8ce9ab15b1a6d58c6aa7a3ceb952d6206814/compiler/rustc_typeck/src/check/upvar.rs#L361)):*
> ```Rust
>     /// ```
>     /// {
>     ///       Place(base: hir_id_s, projections: [], ....) -> (hir_id_L5, ByValue),
> ```

it should be `hir_id_L1` here no? then we need to add a case here fore `hir_id_L5` case? 
___
*[compiler/rustc_typeck/src/check/upvar.rs, line 545 at r5](https://reviewable.io/reviews/sexxi-goose/rust/28#-ML-jNe4DfLUN-kD4gDy:-ML-jNe4DfLUN-kD4gDz:blnjkts) ([raw file](https://github.com/sexxi-goose/rust/blob/ab4b8ce9ab15b1a6d58c6aa7a3ceb952d6206814/compiler/rustc_typeck/src/check/upvar.rs#L545)):*
> ```Rust
> 
>             // Safe to unwrap here, since both are Some(_)
>             let expr_kind_a = self.tcx.hir().get(capture_info_a.expr_id.unwrap());
> ```

Why unwrap here? Maybe move the code here into the `(Some(_), Some(_)) => {}` block above ?
___
*[compiler/rustc_typeck/src/check/upvar.rs, line 555 at r5](https://reviewable.io/reviews/sexxi-goose/rust/28#-ML-j_snBYHLuuHkdnjY:-ML-j_snBYHLuuHkdnjZ:b-oupmpq) ([raw file](https://github.com/sexxi-goose/rust/blob/ab4b8ce9ab15b1a6d58c6aa7a3ceb952d6206814/compiler/rustc_typeck/src/check/upvar.rs#L555)):*
> ```Rust
>                 return capture_info_a;
>             }
>         }
> ```

nit: maybe add an else here? Just a personal preference 


<!-- Sent from Reviewable.io -->

arora-aman reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/sexxi-goose/rust/28)* status: 0 of 33 files reviewed, 12 unresolved discussions (waiting on @arora-aman, @Azhng, and @roxelo)

___
*[compiler/rustc_typeck/src/expr_use_visitor.rs, line 596 at r5](https://reviewable.io/reviews/sexxi-goose/rust/28#-ML-eIwG1gDHIIxfMWaZ:-ML03WiX7Y4YgB82XABs:bdcv6l8) ([raw file](https://github.com/sexxi-goose/rust/blob/ab4b8ce9ab15b1a6d58c6aa7a3ceb952d6206814/compiler/rustc_typeck/src/expr_use_visitor.rs#L596)):*
<details><summary><i>Previously, Azhng (Archer Zhang) wrote…</i></summary><blockquote>

Move this if check to the outer loops? Seems like we are doing redundant checks in the inner tight loops since `var_hir_id` is not updated in the inner loop

Also this double negative seems really strange. Not sure if there's a better way of writing it 🤷‍♀️🤷‍♀️🤷‍♀️
</blockquote></details>

Good point.
___
*[compiler/rustc_typeck/src/check/upvar.rs, line 130 at r5](https://reviewable.io/reviews/sexxi-goose/rust/28#-ML-fh9a8ugBdOs6qPIu:-ML01AzB2s2ETEiyT_0_:b-jul830) ([raw file](https://github.com/sexxi-goose/rust/blob/ab4b8ce9ab15b1a6d58c6aa7a3ceb952d6206814/compiler/rustc_typeck/src/check/upvar.rs#L130)):*
<details><summary><i>Previously, Azhng (Archer Zhang) wrote…</i></summary><blockquote>

So we are keeping the `SG_NEW` env var ?
</blockquote></details>

It's for testing until things are done, will remove before pushing the PR out.
___
*[compiler/rustc_typeck/src/check/upvar.rs, line 153 at r5](https://reviewable.io/reviews/sexxi-goose/rust/28#-ML-hD9t7Wadq6wo5Qsa:-ML01Jhm5VZH5zzy3WLN:bdf7gv8) ([raw file](https://github.com/sexxi-goose/rust/blob/ab4b8ce9ab15b1a6d58c6aa7a3ceb952d6206814/compiler/rustc_typeck/src/check/upvar.rs#L153)):*
<details><summary><i>Previously, Azhng (Archer Zhang) wrote…</i></summary><blockquote>

nit: maybe specified the name of the field name that is being initialized, it's a mix and match here and looks kinda inconsistent
</blockquote></details>

This is a rust thing. If the field being assigned has the same name as the variable being used assign it, then the field name can be skipped in the ctor.  Rust analyzer shows a warning if I put the field name.
___
*[compiler/rustc_typeck/src/check/upvar.rs, line 156 at r5](https://reviewable.io/reviews/sexxi-goose/rust/28#-ML-hMcnB4PAsMIAX-vY:-ML01mFh06SwkWCLuxu0:b-7ropln) ([raw file](https://github.com/sexxi-goose/rust/blob/ab4b8ce9ab15b1a6d58c6aa7a3ceb952d6206814/compiler/rustc_typeck/src/check/upvar.rs#L156)):*
<details><summary><i>Previously, Azhng (Archer Zhang) wrote…</i></summary><blockquote>

ditto 
</blockquote></details>

ditto
___
*[compiler/rustc_typeck/src/check/upvar.rs, line 345 at r5](https://reviewable.io/reviews/sexxi-goose/rust/28#-ML-k8hc8BHofKaVCaEA:-ML01pNc3diR_M9_4s--:bontlfm) ([raw file](https://github.com/sexxi-goose/rust/blob/ab4b8ce9ab15b1a6d58c6aa7a3ceb952d6206814/compiler/rustc_typeck/src/check/upvar.rs#L345)):*
<details><summary><i>Previously, Azhng (Archer Zhang) wrote…</i></summary><blockquote>

Assuming `Point` only has field `x` and `y` here ? 
</blockquote></details>

Yeah doesn't really affect it AGA
___
*[compiler/rustc_typeck/src/check/upvar.rs, line 361 at r5](https://reviewable.io/reviews/sexxi-goose/rust/28#-ML-kJvn8u6FjJgGXENE:-ML01zzYAAzKd0gi3j2_:b5rbdlr) ([raw file](https://github.com/sexxi-goose/rust/blob/ab4b8ce9ab15b1a6d58c6aa7a3ceb952d6206814/compiler/rustc_typeck/src/check/upvar.rs#L361)):*
<details><summary><i>Previously, Azhng (Archer Zhang) wrote…</i></summary><blockquote>

it should be `hir_id_L1` here no? then we need to add a case here fore `hir_id_L5` case? 
</blockquote></details>

No, InferBorrowKind handles the different CaptureKinds for the same place. `hir_id_L5` has the highest prefernece since it results in ByValue.
___
*[compiler/rustc_typeck/src/check/upvar.rs, line 545 at r5](https://reviewable.io/reviews/sexxi-goose/rust/28#-ML-jNe4DfLUN-kD4gDy:-ML02lCuAw5MKkFeXnBN:bdnhvfg) ([raw file](https://github.com/sexxi-goose/rust/blob/ab4b8ce9ab15b1a6d58c6aa7a3ceb952d6206814/compiler/rustc_typeck/src/check/upvar.rs#L545)):*
<details><summary><i>Previously, Azhng (Archer Zhang) wrote…</i></summary><blockquote>

Why unwrap here? Maybe move the code here into the `(Some(_), Some(_)) => {}` block above ?
</blockquote></details>

I was avoiding nesting, but sure.
___
*[compiler/rustc_typeck/src/check/upvar.rs, line 555 at r5](https://reviewable.io/reviews/sexxi-goose/rust/28#-ML-j_snBYHLuuHkdnjY:-ML02p89-zzPkQBwOVvq:b5loqya) ([raw file](https://github.com/sexxi-goose/rust/blob/ab4b8ce9ab15b1a6d58c6aa7a3ceb952d6206814/compiler/rustc_typeck/src/check/upvar.rs#L555)):*
<details><summary><i>Previously, Azhng (Archer Zhang) wrote…</i></summary><blockquote>

nit: maybe add an else here? Just a personal preference 
</blockquote></details>

I think it adds a lot of nesting


<!-- Sent from Reviewable.io -->

arora-aman commented:
___
*[compiler/rustc_typeck/src/check/upvar.rs, line 345 at r5](https://reviewable.io/reviews/sexxi-goose/rust/28#-ML-k8hc8BHofKaVCaEA:-ML08nao6jNZ7CtwSe1H:bl0dd23) ([raw file](https://github.com/sexxi-goose/rust/blob/ab4b8ce9ab15b1a6d58c6aa7a3ceb952d6206814/compiler/rustc_typeck/src/check/upvar.rs#L345)):*
<details><summary><i>Previously, arora-aman (Aman Arora) wrote…</i></summary><blockquote>

Yeah doesn't really affect it AGA
</blockquote></details>

~AGA~, i'll still add the declaration for Point


<!-- Sent from Reviewable.io -->

arora-aman committed:
Handle `let _ = x` patterns in closure liveness analysis

arora-aman commented:
@Azhng updated

arora-aman reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/sexxi-goose/rust/28)* status: 0 of 34 files reviewed, 9 unresolved discussions (waiting on @Azhng and @roxelo)

___
*[compiler/rustc_mir/src/borrow_check/diagnostics/mod.rs, line 386 at r5](https://reviewable.io/reviews/sexxi-goose/rust/28#-ML-bHgb5V7kQh5pfti3:-ML6OZjFBrShewTflwp_:b-896fix) ([raw file](https://github.com/sexxi-goose/rust/blob/ab4b8ce9ab15b1a6d58c6aa7a3ceb952d6206814/compiler/rustc_mir/src/borrow_check/diagnostics/mod.rs#L386)):*
<details><summary><i>Previously, Azhng (Archer Zhang) wrote…</i></summary><blockquote>

The comment here need to be updated since we are reading from `closure_captures` instead of `upvars_mentioned`
</blockquote></details>

Done.
___
*[compiler/rustc_typeck/src/expr_use_visitor.rs, line 596 at r5](https://reviewable.io/reviews/sexxi-goose/rust/28#-ML-eIwG1gDHIIxfMWaZ:-ML6Ob2E9ELHIW3ihpUH:b-896fix) ([raw file](https://github.com/sexxi-goose/rust/blob/ab4b8ce9ab15b1a6d58c6aa7a3ceb952d6206814/compiler/rustc_typeck/src/expr_use_visitor.rs#L596)):*
<details><summary><i>Previously, arora-aman (Aman Arora) wrote…</i></summary><blockquote>

Good point.
</blockquote></details>

Done.
___
*[compiler/rustc_typeck/src/check/upvar.rs, line 61 at r3](https://reviewable.io/reviews/sexxi-goose/rust/28#-MKMSR7YBP9hf0mtSKJ6-r3-61:-ML6ObaFCWwJm_7wpPtC:b-896fix) ([raw file](https://github.com/sexxi-goose/rust/blob/9007f13ba4f84a4b64ee3cb1599008574364aeff/compiler/rustc_typeck/src/check/upvar.rs#L61)):*
<details><summary><i>Previously, arora-aman (Aman Arora) wrote…</i></summary><blockquote>

Updated the last commit
</blockquote></details>

Done.
___
*[compiler/rustc_typeck/src/check/upvar.rs, line 53 at r5](https://reviewable.io/reviews/sexxi-goose/rust/28#-ML-f0p79ZtaSF2ZIbx7:-ML6OePXBOI9_bUTKeph:b-896fix) ([raw file](https://github.com/sexxi-goose/rust/blob/ab4b8ce9ab15b1a6d58c6aa7a3ceb952d6206814/compiler/rustc_typeck/src/check/upvar.rs#L53)):*
<details><summary><i>Previously, Azhng (Archer Zhang) wrote…</i></summary><blockquote>

nit: the order of comment is inconsistent with the definition of the enum 
</blockquote></details>

Done.
___
*[compiler/rustc_typeck/src/check/upvar.rs, line 153 at r5](https://reviewable.io/reviews/sexxi-goose/rust/28#-ML-hD9t7Wadq6wo5Qsa:-ML6OvG2-ovc92usSOuG:bb6un1j) ([raw file](https://github.com/sexxi-goose/rust/blob/ab4b8ce9ab15b1a6d58c6aa7a3ceb952d6206814/compiler/rustc_typeck/src/check/upvar.rs#L153)):*
<details><summary><i>Previously, arora-aman (Aman Arora) wrote…</i></summary><blockquote>

This is a rust thing. If the field being assigned has the same name as the variable being used assign it, then the field name can be skipped in the ctor.  Rust analyzer shows a warning if I put the field name.
</blockquote></details>

Done
___
*[compiler/rustc_typeck/src/check/upvar.rs, line 156 at r5](https://reviewable.io/reviews/sexxi-goose/rust/28#-ML-hMcnB4PAsMIAX-vY:-ML6OhTUEAPC46zNymzl:bb6un1j) ([raw file](https://github.com/sexxi-goose/rust/blob/ab4b8ce9ab15b1a6d58c6aa7a3ceb952d6206814/compiler/rustc_typeck/src/check/upvar.rs#L156)):*
<details><summary><i>Previously, arora-aman (Aman Arora) wrote…</i></summary><blockquote>

ditto
</blockquote></details>

Done
___
*[compiler/rustc_typeck/src/check/upvar.rs, line 345 at r5](https://reviewable.io/reviews/sexxi-goose/rust/28#-ML-k8hc8BHofKaVCaEA:-ML6OkSV2FSodRlbH3_L:b-896fix) ([raw file](https://github.com/sexxi-goose/rust/blob/ab4b8ce9ab15b1a6d58c6aa7a3ceb952d6206814/compiler/rustc_typeck/src/check/upvar.rs#L345)):*
<details><summary><i>Previously, arora-aman (Aman Arora) wrote…</i></summary><blockquote>

~AGA~, i'll still add the declaration for Point
</blockquote></details>

Done.
___
*[compiler/rustc_typeck/src/check/upvar.rs, line 545 at r5](https://reviewable.io/reviews/sexxi-goose/rust/28#-ML-jNe4DfLUN-kD4gDy:-ML6OlO97ampdAHXCXlf:b-896fix) ([raw file](https://github.com/sexxi-goose/rust/blob/ab4b8ce9ab15b1a6d58c6aa7a3ceb952d6206814/compiler/rustc_typeck/src/check/upvar.rs#L545)):*
<details><summary><i>Previously, arora-aman (Aman Arora) wrote…</i></summary><blockquote>

I was avoiding nesting, but sure.
</blockquote></details>

Done.
___
*[compiler/rustc_typeck/src/check/upvar.rs, line 555 at r5](https://reviewable.io/reviews/sexxi-goose/rust/28#-ML-j_snBYHLuuHkdnjY:-ML6OmK9EIPD6ymvtNFZ:bb6un1j) ([raw file](https://github.com/sexxi-goose/rust/blob/ab4b8ce9ab15b1a6d58c6aa7a3ceb952d6206814/compiler/rustc_typeck/src/check/upvar.rs#L555)):*
<details><summary><i>Previously, arora-aman (Aman Arora) wrote…</i></summary><blockquote>

I think it adds a lot of nesting
</blockquote></details>

Done


<!-- Sent from Reviewable.io -->

arora-aman committed:
More pattern testcases

arora-aman committed:
Remove local testing env var

Azhng reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

> *[Reviewable](https://reviewable.io/reviews/sexxi-goose/rust/28#-:-MLRFSxu5GW9l-Lv_NWP:bg7igkv)* status: 0 of 35 files reviewed, 2 unresolved discussions (waiting on @Azhng and @roxelo)



<!-- Sent from Reviewable.io -->

arora-aman committed:
Address review comments

arora-aman committed:
Address review comments 2

arora-aman committed:
Add test for capturing enums

arora-aman committed:
Log closure as well

arora-aman committed:
Fix case when ExprUseVisitor is called after typeck writeback

Clippy uses `ExprUseVisitor` and atleast in some cases it runs
after writeback.

We currently don't writeback the min_capture results of closure
capture analysis since no place within the compiler itself uses it.

In the short term to fix clippy we add a fallback when walking captures
of a closure to check if closure_capture analysis has any entries in it.

Writeback for closure_min_captures will be implemented in
rust-lang/project-rfc-2229#18

arora-aman committed:
Ignore doctest for capture analysis examples
